### PR TITLE
fix file/url path handling for nbviewer and server launch url

### PR DESCRIPTION
**Pull request by bitnik**
with https://github.com/jupyterhub/binderhub/commit/5c993cf4f21d1c2d0274aa18d23259baa213cbae I tried to make js code easier to read and follow. 

https://github.com/jupyterhub/binderhub/commit/c03211f13a4cce7308bd06b1e8c1024fbf7e5199 fixes https://github.com/jupyterhub/binderhub/issues/456 and https://github.com/jupyterhub/binderhub/issues/622

https://github.com/jupyterhub/binderhub/commit/23e94ede3d9230426236c9c4bb9dfc656ef13236 updates loading page test and adds more cases. There I used `berndweiss/gesis-meta-analysis-2018` repo for some cases. If you want, I can change it (and also `minrk/ligo-binder` repo) with https://github.com/binderhub-ci-repos/requirements but then some changes are needed there.

### CODE REVIEW

File: binderhub/main.py [hunk 1]
```diff
@@ -25,12 +25,9 @@ class ParameterizedMainHandler(BaseHandler):
...
 
     def get(self, provider_prefix, _unescaped_spec):
-        # re-extract spec from request.path
-        # get the original, raw spec, without tornado's unquoting
-        # this is needed because tornado converts 'foo%2Fbar/ref' to 'foo/bar/ref'
         prefix = '/v2/' + provider_prefix
-        idx = self.request.path.index(prefix)
-        spec = self.request.path[idx + len(prefix) + 1:]
+        spec = self.get_spec_from_request(prefix)
+        spec = spec.rstrip("/")
```
Comment (bitnik) [inline binderhub/main.py]: `spec = spec.rstrip("/")` also fix for https://github.com/jupyterhub/binderhub/issues/622.

File: binderhub/main.py [hunk 2]
```diff
@@ -44,12 +41,22 @@ def get(self, provider_prefix, _unescaped_spec):
...
             raise web.HTTPError(400, str(e))
 
+        provider_spec = f'{provider_prefix}/{spec}'
+        nbviewer_url = None
+        if provider_prefix == "gh":
+            # we can only produce an nbviewer URL for github right now
+            nbviewer_url = 'https://nbviewer.jupyter.org/github'
+            org, repo, ref = spec.split('/', 2)
+            # NOTE: tornado escapes query arguments too -> notebooks%2Findex.ipynb becomes notebooks/index.ipynb
+            filepath = self.get_argument('filepath', '').lstrip('/')
```
Comment (bitnik) [inline binderhub/main.py]: `filepath = self.get_argument('filepath', '').lstrip('/')` fixes #622

File: binderhub/main.py (hunk 1 above)
Comment (bitnik) [reply binderhub/main.py]: that line also does the same for nbviewer as done for build in https://github.com/jupyterhub/binderhub/blob/master/binderhub/builder.py#L205

File: binderhub/static/js/index.js [hunk 3]
```diff
@@ -90,36 +90,36 @@ Image.prototype.close = function() {
...
 };
-
-Image.prototype.launch = function(url, token, filepath, pathType) {
+Image.prototype.launch = function(url, token, path, pathType) {
     // redirect a user to a running server with a token
-    if (filepath) {
+    if (path) {
       // strip trailing /
       url = url.replace(/\/$/, '');
+      path = decodeURIComponent(path);
```
Comment (minrk) [inline binderhub/static/js/index.js]: Need to be careful with decoding/encoding path. There are two patch kinds and two path sources, which will have different encoding:

- from url param (encoded)
- filepath from form (not encoded)
- urlpath from form (encoded)

Need to make sure that the input path is always treated correctly.

Calling decode here means that encode also needs to be called below in order to preserve the right value.

**Comment by bitnik** [comment]
@minrk I deleted the line with decode.

But if I understood your comment correctly, there is probably still sth wrong. [This part of the js](https://github.com/jupyterhub/binderhub/blob/master/binderhub/static/js/index.js#L408-L418) returns the decoded path value. Maybe we should get the raw query arguments in python code from `request.query`. What do you think?

**Comment by minrk** [comment]
It's tricky! Let's try to define APIs carefully:

How about this:

- filepaths are not encoded because they should reflect the logical path in the repo (e.g. `/path to/file.ipynb`)
- urlpaths are encoded because they are URLs (e.g. `/notebooks/path%20/file.ipynb`)

When coming in from query params, both are encoded by necessity (`?filepath=/path%20file.ipynb&urlpath=/notebooks/path%20file.ipynb`).

I think that means that in general, our code should act as if filepaths are not encoded and urlpaths are, and when filepath is coming from the url param, we should decode it immediately first, so that it can be handled appropriately throughout, the same as if it had been typed in the form.

That means in `build`, taking a path, *filepath* must call `encodeURIComponent` prior to building the URL and *urlpath* must NOT encode the URI.

**Comment by bitnik** [comment]
is it still not correct? @minrk

**Comment by minrk** [comment]
I think it's all set now, thanks!

**Comment by bitnik** [comment]
thank you @minrk!

**Pull request merged**